cpp/src/parquet/encoding.cc

```diff
@@ -2062,6 +2062,184 @@ class DictByteArrayDecoderImpl : public DictDecoderImpl<ByteArrayType>,
+    if (values_current_block_ == values_per_block_) {
+      FlushBlock();
+    }
+  }
+
+  if (values_current_block_ != 0) {
+    FlushBlock();
+  }
+}
+
+template <typename DType>
+void DeltaBitPackEncoder<DType>::FlushBlock() {
+  if (values_current_block_ == 0) {
+    return;
+  }
+
+  const auto min_delta =
+      static_cast<int64_t>(*std::min_element(deltas_.begin(), deltas_.end()));
+  DCHECK(bit_writer_.PutZigZagVlqInt(min_delta));
+
+  uint8_t* bit_width_offsets = bit_writer_.GetNextBytePtr(mini_blocks_per_block_);
+  DCHECK(bit_width_offsets != NULL);
+
+  for (uint32_t i = 0; i < mini_blocks_per_block_; i++) {
+    const uint32_t n = std::min(values_per_mini_block_, values_current_block_);
+    if (n == 0) {
+      DCHECK(
+          bit_writer_.PutAlignedOffset<uint32_t>(bit_width_offsets + i, uint32_t(1), 1));
+      continue;
+    }
+
+    const uint32_t start = i * values_per_mini_block_;
+    const auto max_delta = static_cast<int64_t>(
+        *std::max_element(deltas_.begin() + start, deltas_.begin() + start + n));
+
+    const uint32_t num_bits = bit_util::NumRequiredBits(max_delta - min_delta);
+    DCHECK(bit_writer_.PutAlignedOffset<uint32_t>(bit_width_offsets + i, num_bits, 1));
+
+    for (uint64_t j = start; j < start + n; j++) {
+      DCHECK(
+          bit_writer_.PutValue(static_cast<uint64_t>(deltas_[j] - min_delta), num_bits));
+    }
+    for (uint64_t j = n; j < values_per_mini_block_; j++) {
+      DCHECK(bit_writer_.PutValue(0, num_bits));
+    }
+    values_current_block_ -= n;
+  }
+  DCHECK_EQ(values_current_block_, 0);
+
+  PARQUET_THROW_NOT_OK(sink_.Append(bit_writer_.buffer(), bit_writer_.bytes_written()));
+  bit_writer_.Clear();
+}
+
+template <typename DType>
+std::shared_ptr<Buffer> DeltaBitPackEncoder<DType>::FlushValues() {
+  std::shared_ptr<ResizableBuffer> header_buffer = AllocateBuffer(pool_, 32);
+  ::arrow::bit_util::BitWriter header_writer(header_buffer->mutable_data(),
+                                             static_cast<int>(header_buffer->size()));
+  if (!header_writer.PutVlqInt(values_per_block_) ||
+      !header_writer.PutVlqInt(mini_blocks_per_block_) ||
+      !header_writer.PutVlqInt(total_value_count_) ||
+      !header_writer.PutZigZagVlqInt(first_value_)) {
+    throw ParquetException("cannot write");
+  }
+  header_writer.Flush(false);
+
+  ::arrow::BufferBuilder sink;
+  PARQUET_THROW_NOT_OK(
+      sink.Append(header_writer.buffer(), header_writer.bytes_written()));
+  header_writer.Clear();
+
+  std::shared_ptr<Buffer> bits_buffer;
+  PARQUET_THROW_NOT_OK(sink_.Finish(&bits_buffer, true));
+
+  std::shared_ptr<Buffer> buffer;
+  PARQUET_THROW_NOT_OK(sink.Append(bits_buffer->mutable_data(), bits_buffer->size()));
+  PARQUET_THROW_NOT_OK(sink.Finish(&buffer, true));
+  return buffer;
+}
+
+template <>
+void DeltaBitPackEncoder<Int32Type>::Put(const ::arrow::Array& values) {
+  auto src = reinterpret_cast<int32_t*>(values.data()->buffers[0]->mutable_data());
```

Comment:
Hmm, why `buffers[0]`? That would be the null bitmap, no?

Probably something like `values.data()->GetValues<int32_t>(1)` instead.

Reply:
Indeed, changed.